Use the CMS_THREAD_SAFE macro instead of [[cms::thread_safe]]

fwyzard opened the pull request:
The Intel Compiler does not seem to support attributes with a namespace:

> warning #3924: attribute namespace "cms" is unrecognized

fwyzard committed:
Do not define attributes in the cms namespace when using ICC

The Intel Compiler does not seem to support attributes with a
namespace:

> warning #3924: attribute namespace "cms" is unrecognized

fwyzard committed:
Use the CMS_THREAD_SAFE macro instead of [[cms::thread_safe]]

fwyzard commented:
@cmsbuild, please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21480/2224

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for master.

It involves the following packages:

FWCore/Common
FWCore/Framework
FWCore/MessageLogger
FWCore/MessageService
FWCore/ParameterSet
FWCore/PluginManager
FWCore/Utilities
Utilities/XrdAdaptor


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @felicepantaleo, @ebrondol, @wddgit, @makortel this is something you requested to watch as well.
@davidlange6, @slava77 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar (and backports should be raised in the release meeting by the corresponding L2)

fwyzard committed:
Use the CMS_THREAD_SAFE and CMS_THREAD_GUARD(...) macros

Use the `CMS_THREAD_SAFE` and `CMS_THREAD_GUARD(...)` macros
instead of using directly the `[[cms::thread_safe]]` and
`[[cms::thread_guard(...)]]` attributes.

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
@cmsbuild, please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21480/2231

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #21480 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @lveldere, @arunhep, @ssekmen, @cerminar, @monttj, @cmsbuild, @franzoni, @ghellwig, @mdhildreth, @slava77, @lpernie can you please check and sign again.

cmsbuild commented:
-1

Tested at: 4bbf2d404e25139aaddcf4f8fe17f9271b258fca


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/f493624b3018543865bbf04bb8a48c5dae44bc82
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24733/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24733/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24733/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found an error when building: 
 
 <pre>>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-27-2300/src/SimG4Core/Watcher/test/IntSimProducer.cc 
Entering library rule at src/SimG4Core/CustomPhysics/plugins
>> Compiling  /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-27-2300/src/SimG4Core/CustomPhysics/src/CustomPDGParser.cc 
>> Compiling  /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-27-2300/src/SimG4Core/CustomPhysics/src/CMSDarkPairProductionProcess.cc 
>> Compiling  /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-27-2300/src/SimG4Core/CustomPhysics/src/CustomPhysicsListSS.cc 
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-27-2300/src/SimG4Core/CustomPhysics/src/CMSDarkPairProductionProcess.cc:8:77: fatal error: SimG4Core/CustomPhysics/interface/CMSDarkPairProductionProcess.hh: No such file or directory
 #include "SimG4Core/CustomPhysics/interface/CMSDarkPairProductionProcess.hh"
                                                                             ^
compilation terminated.
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-27-2300/src/SimG4Core/CustomPhysics/src/CMSDarkPairProductionProcess.cc:8:77: fatal error: SimG4Core/CustomPhysics/interface/CMSDarkPairProductionProcess.hh: No such file or directory
 #include "SimG4Core/CustomPhysics/interface/CMSDarkPairProductionProcess.hh"
</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/f493624b3018543865bbf04bb8a48c5dae44bc82
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24733/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24733/git-merge-result

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

fwyzard commented:
The error is unrelated to these changes.

perrotta commented:
please test
(Last CMSSW_10_0_X_2017-11-28-2300 IB should be fixed now)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 4bbf2d404e25139aaddcf4f8fe17f9271b258fca
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24741/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/bd23f8cee69cf9c2a2bfcf496488b985d646463c
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24741/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24741/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24741/summary.html

There are some workflows for which there are errors in the baseline:
 10824.0 step 5
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2605338
 - DQMHistoTests: Total failures: 21
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2605146
 - DQMHistoTests: Total skipped: 171
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.920000000173 KiB( 22 files compared)
 - Checked 106 log files, 8 edm output root files, 26 DQM output files

Dr15Jones commented:
+1

civanch commented:
+1

davidlange6 commented:
Your PR is unmergeable. Please have a look and possibly rebase it.

fwyzard commented:
rebased

fwyzard commented:
@cmsbuild, please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21480/2377

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/PR-21480/2377/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/PR-21480/2377/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

fwyzard commented:
The suggested changes for the code checks
```
diff --git a/FWCore/MessageLogger/src/ErrorObj.cc b/FWCore/MessageLogger/src/ErrorObj.cc
index d58228c7a89..8bfb5d2e981 100644
--- a/FWCore/MessageLogger/src/ErrorObj.cc
+++ b/FWCore/MessageLogger/src/ErrorObj.cc
@@ -246,7 +246,7 @@ void ErrorObj::set( const ELseverityLevel & sev, const ELstring & id )  {
   clear();
 
   myTimestamp = time( nullptr );
-  mySerial = ++ ourSerial;
+  mySerial = ++ mySerial;
 
   setID( id );
   setSeverity( sev );
diff --git a/PhysicsTools/MVAComputer/src/AtomicId.cc b/PhysicsTools/MVAComputer/src/AtomicId.cc
index da07cd65a00..7d1bd031544 100644
--- a/PhysicsTools/MVAComputer/src/AtomicId.cc
+++ b/PhysicsTools/MVAComputer/src/AtomicId.cc
@@ -59,7 +59,7 @@ namespace PhysicsTools {
 static IdCache &getAtomicIdCache()
 {
 	CMS_THREAD_SAFE static IdCache atomicIdCache;
-	return atomicIdCache;
+	return getAtomicIdCache;
 }
 
 const char *AtomicId::lookup(const char *string) throw()
```
look wrong !

davidlange6 commented:
indeed - lets see if its reproducible

> On Nov 30, 2017, at 7:14 AM, Andrea Bocci <notifications@github.com> wrote:
> 
> The suggested changes for the code checks
> 
> diff --git a/FWCore/MessageLogger/src/ErrorObj.cc b/FWCore/MessageLogger/src/ErrorObj.cc
> index d58228c7a89..8bfb5d2e981 100644
> --- a/FWCore/MessageLogger/src/ErrorObj.cc
> +++ b/FWCore/MessageLogger/src/ErrorObj.cc
> @@ -246,7 +246,7 @@ void ErrorObj::set( const ELseverityLevel & sev, const ELstring & id )  {
>    clear();
>  
>    myTimestamp = time( nullptr );
> -  mySerial = ++ ourSerial;
> +  mySerial = ++ mySerial;
>  
>    setID( id );
>    setSeverity( sev );
> diff --git a/PhysicsTools/MVAComputer/src/AtomicId.cc b/PhysicsTools/MVAComputer/src/AtomicId.cc
> index da07cd65a00..7d1bd031544 100644
> --- a/PhysicsTools/MVAComputer/src/AtomicId.cc
> +++ b/PhysicsTools/MVAComputer/src/AtomicId.cc
> @@ -59,7 +59,7 @@ namespace PhysicsTools {
>  static IdCache &getAtomicIdCache()
>  {
>  	CMS_THREAD_SAFE static IdCache atomicIdCache;
> -	return atomicIdCache;
> +	return getAtomicIdCache;
>  }
>  
>  const char *AtomicId::lookup(const char *string) throw()
> 
> look wrong !
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

davidlange6 commented:
code-checks

> On Nov 30, 2017, at 7:14 AM, Andrea Bocci <notifications@github.com> wrote:
> 
> The suggested changes for the code checks
> 
> diff --git a/FWCore/MessageLogger/src/ErrorObj.cc b/FWCore/MessageLogger/src/ErrorObj.cc
> index d58228c7a89..8bfb5d2e981 100644
> --- a/FWCore/MessageLogger/src/ErrorObj.cc
> +++ b/FWCore/MessageLogger/src/ErrorObj.cc
> @@ -246,7 +246,7 @@ void ErrorObj::set( const ELseverityLevel & sev, const ELstring & id )  {
>    clear();
>  
>    myTimestamp = time( nullptr );
> -  mySerial = ++ ourSerial;
> +  mySerial = ++ mySerial;
>  
>    setID( id );
>    setSeverity( sev );
> diff --git a/PhysicsTools/MVAComputer/src/AtomicId.cc b/PhysicsTools/MVAComputer/src/AtomicId.cc
> index da07cd65a00..7d1bd031544 100644
> --- a/PhysicsTools/MVAComputer/src/AtomicId.cc
> +++ b/PhysicsTools/MVAComputer/src/AtomicId.cc
> @@ -59,7 +59,7 @@ namespace PhysicsTools {
>  static IdCache &getAtomicIdCache()
>  {
>  	CMS_THREAD_SAFE static IdCache atomicIdCache;
> -	return atomicIdCache;
> +	return getAtomicIdCache;
>  }
>  
>  const char *AtomicId::lookup(const char *string) throw()
> 
> look wrong !
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21480/2380

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/PR-21480/2380/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/PR-21480/2380/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

fwyzard commented:
Seems it is reproducibly wrong.

davidlange6 commented:
probably its because its using a broken IB

> On Nov 30, 2017, at 8:07 AM, Andrea Bocci <notifications@github.com> wrote:
> 
> Seems it is reproducibly wrong.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
code-checks

fwyzard commented:
Can you please stop asking for useless code checks, and just merge this, please ?

slava77 commented:
On 11/30/17 10:19 AM, Andrea Bocci wrote:
> Can you please stop asking for useless code checks, and just merge this,
> please ?
> 

current strategy is that the code-checks is blocking.
If this is merged, the next time these files are modified with
functional changes
the code-checks will fail.
IMHO, the code-checker should be fixed before this can be merged.


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/21480#issuecomment-348275262>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbsJqiCXmsH2Bp3o2-NUujNXLutXzks5s7vHFgaJpZM4Qt4wV>.
>

fwyzard commented:
@slava77 I understand what you say, but I disagree with your conclusion: the lines affected by the code checks are already there before this PR, so any changes affecting those files will run into the same problem, irrespective of this PR being merged or not.

My point is that this PR does not create new code check issues, and is only a trivial clean up, so I'd rather have it merged sooner than later, and especially before any other PR that may require it to be rebased again.

slava77 commented:
On 11/30/17 1:46 PM, Andrea Bocci wrote:
> @slava77 <https://github.com/slava77> I understand what you say, but I
> disagree with your conclusion: the lines affected by the code checks are
> already there before this PR, so any changes affecting those files will
> run into the same problem, irrespective of this PR being merged or not.
> 
> My point is that this PR does not create new code check issues, and is
> only a trivial clean up, so I'd rather have it merged sooner than later,
> and especially before any other PR that may require it to be rebased again.

in both cases the code-checks patch points to a variable which
declaration has "CMS_THREAD_SAFE" prepended, each from this PR edits.
So, it seemed to me that there is a connection

Are you saying that the code-checks fails and suggests the same nonsense
changes  without your changes?



> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/21480#issuecomment-348331045>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbk4ji7wQBB5YkZtHZPlzx-QT42KXks5s7yIYgaJpZM4Qt4wV>.
>

fwyzard commented:
I couldn't say.
Running `scram build code-checks` with this branch on top of CMSSW_10_0_X_2017-11-30-1100 does not spot any code problems anyway, so I cannot check if it would find any problems without the changes...

fwyzard commented:
@cmsbuild, please check ...

davidlange6 commented:
code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21480/2390

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #21480 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @lveldere, @arunhep, @ssekmen, @cerminar, @monttj, @cmsbuild, @franzoni, @ghellwig, @mdhildreth, @slava77, @lpernie can you please check and sign again.

davidlange6 commented:
so indeed the problem was the underlying IB not the code checker..

ghellwig commented on Calibration/Tools/interface/EcalRingCalibrationTools.h:
```diff
@@ -48,7 +49,7 @@ class EcalRingCalibrationTools
   static std::atomic<bool> isInitializedFromGeometry_;
 
-  [[cms::thread_guard("isInitializedFromGeometry_")]]
+  CMS_THREAD_GUARD("isInitializedFromGeometry_")
```
@fwyzard maybe, I am not experienced enough with macros, but given the definition of `CMS_THREAD_GUARD`, I would rather translate it to this statement:
```
CMS_THREAD_GUARD(isInitializedFromGeometry_)
```

fwyzard replied:
Good point... we are already using both syntaxes e.g. in [[1]] and [[2]].
Let me figure out which one is correct and fix the others...

[1]: https://github.com/cms-sw/cmssw/blob/master/DataFormats/Common/interface/DetSetVectorNew.h#L111
[2]: https://github.com/cms-sw/cmssw/blob/master/DataFormats/PatCandidates/interface/PackedCandidate.h#L727

fwyzard replied:
I think the correct syntax should be:
```c++
#define CMS_THREAD_GUARD(_var_) [[cms::thread_guard(#_var_)]]
...
CMS_THREAD_GUARD(isInitializedFromGeometry_)
static short endcapRingIndex_[EEDetId::IX_MAX][EEDetId::IY_MAX];
```

fwyzard replied:
The present definition
```c++
#define CMS_THREAD_GUARD(_var_) [[cms::thread_guard("#_var_")]]
```
expands to
```
[[cms::thread_guard("#_var_")]]
```
no matter the argument of the macro - which I doubt was the intended behaviour.

fwyzard committed:
Fix the definition and usage of CMS_THREAD_GUARD

fwyzard commented:
@cmsbuild, please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21480/2391

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #21480 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @lveldere, @arunhep, @ssekmen, @cerminar, @monttj, @cmsbuild, @franzoni, @ghellwig, @mdhildreth, @slava77, @lpernie can you please check and sign again.

ghellwig commented:
+1
provided jenkins agrees

Dr15Jones commented:
+1

cmsbuild commented:
+1
Tested at: 69ed6b8df6b9a849b283bbe6966a049a063d7f07
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24903/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/9b389656fc7abd70a8507259b1eb6ee57a00d319
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24903/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24903/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21480/24903/summary.html

There are some workflows for which there are errors in the baseline:
 10824.0 step 5
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2715851
 - DQMHistoTests: Total failures: 21
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2715654
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.17999999998 KiB( 23 files compared)
 - Checked 111 log files, 8 edm output root files, 27 DQM output files

civanch commented:
+1

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/21480	69ed6b8
- technical update
- jenkins tests pass

davidlange6 commented:
merge
